keel-core/src/main/kotlin/com/netflix/spinnaker/keel/veto/unhappy/UnhappyVeto.kt

```diff
@@ -56,20 +60,29 @@ class UnhappyVeto(
     }
 
     val wait = waitingTime(resource)
-    val vetoStatus = unhappyVetoRepository.getOrCreateVetoStatus(resourceId, application, wait)
-    // allow for a check every [waitingTime] even if the resource is unhappy
-    if (vetoStatus.shouldRecheck) {
-      log.debug("Marking resource $resourceId unhappy for $wait, but allowing resource check.")
-      unhappyVetoRepository.markUnhappyForWaitingTime(resourceId, application, wait)
-      return allowedResponse()
-    }
+    val recheckTime = unhappyVetoRepository.getRecheckTime(resourceId)
 
-    if (vetoStatus.shouldSkip) {
+    /**
```

Comment:
I consolidated all the business logic for unhappy vetos here so it's easy to reason about. I added a comment describing the algorithm which hopefully also helps others understand the code.